### DISCUSSION TITLE
Filter or hide some configuration options in the CLI

### DISCUSSION THREAD

**Discussion by DGuhr** [keycloak/keycloak#10499] [Archive]
### Current situation:
The first class config is growing, as is the count of configuration options in the CLI and overall. To not let this get messy, and to provide an easy way to find the specific configuration, we should provide a way to better find the options you search. 

### Proposal
Make filtering possible using e.g. `kc.sh --help <category>` or `kc.sh --help --<part-of-configuration-key>`

so when i only want to see the database related configuration options, I'd use e.g. `kc.sh start --help database`

or when I only want to see configuration options beginning with `https` I'd use `kc.sh start --help https`

where the term to filter for is not case sensitive.

Considerations:
- There may be other options, such as an "expert mode" and a "default mode" only showing a subset. I'd be in favor of using filters, but that's just me ;)
- decouple category / key search by providing a `-c` and `-k`? So `kc.sh start --help --category=HTTP/TLS` would return all HTTP/TLS related stuff from the category, `kc.sh start --help --key=http` would return all keys where http occurs no matter the category?
- alternative using e.g. `kc.sh config search ...?` so have another toplvl command?

**Comment by bs-matil** [comment]
a different way to do this is using subcommands. I wonder if keycloak would profit from it. 

e.g kc.sh database --some--setting value/ --help

**Comment by stianst** [comment]
--help category works for me. Don't think that should be mixed with search though.

If we want a help with filters it should be a single option, something like --help-search

I can also see that we would want some advanced options, and would like to see that concept baked into this proposal, so we can address the UX in more holistically.

Would also like to know what impact it would have on --help. Is that still supported or is a category needed. How does a user find out about --help category if the initial page is too long, etc.

**Reply by stianst** [reply]
This proposal also has some impact on the proposal from @pedroigor to combine build/runtime options, and making build an optional thing. As that would result in more options to start, and also the need to be able to discover through the CLI what options can be optimised by a build. Perhaps the latter is just using build --help to se those, rather than a marker in start --help
